File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-213846-32199/analysis/acceptance-project/src/main/java/customerimporter/CUIT.java
package customerimporter;

import java.util.List;
import java.util.function.Supplier;

public class CUIT extends Identification {

    public static final String TYPE = "C";

    // "23-25666777-9" size 13
    private static final int MINIMUM_SIZE = 12;
    private static final int MAXIMUM_SIZE = 13;
    private static final int FIRST_DASH_INDEX = 2;
    private static final List<String> VALID_PREFIXES = List.of("20", "23", "24", "25", "26", "27", "30", "33", "34");
    private static final CodeValidator validator = CodeValidator.failingWith(invalidCUITNumberErrorDescription());

    private final String number;

    // instance creation

    public static CUIT from(String aNumber) {
        assertIsValidNumber(aNumber);
        return new CUIT(aNumber);
    }

    // error messages

    public static String invalidCUITNumberErrorDescription() {
        return "Invalid CUIT number";
    }

    // asserting - private

    private static void assertIsValidNumber(String aNumber) {
        validator.assertSizeBetween(aNumber, MINIMUM_SIZE, MAXIMUM_SIZE);
        validator.assertCharAtIs(aNumber, FIRST_DASH_INDEX, '-');
        validator.assertCharAtIs(aNumber, aNumber.length() - 2, '-');
        validator.assertIsOneOf(prefixOf(aNumber), VALID_PREFIXES);
        validator.assertAllDigits(verificationDigitOf(aNumber));
        validator.assertAllDigits(identificationNumberOf(aNumber));
    }

    private static String identificationNumberOf(String aNumber) {
        return aNumber.substring(FIRST_DASH_INDEX + 1, aNumber.length() - 2);
    }

    private static String prefixOf(String aNumber) {
        return aNumber.substring(0, FIRST_DASH_INDEX);
    }

    private static String verificationDigitOf(String aNumber) {
        return aNumber.substring(aNumber.length() - 1);
    }

    // initialization

    private CUIT(String aNumber) {
        number = aNumber;
    }

    // identification

    public String number() {
        return number;
    }

    public String type() {
        return TYPE;
    }

    // testing

    public boolean isCUIT() {
        return true;
    }

    public boolean isDNI() {
        return false;
    }

    // numbers

    public Object cuitNumberIfNone(Supplier<Object> aNoneBlock) {
        return number;
    }

    public Object dniNumberIfNone(Supplier<Object> aNoneBlock) {
        return aNoneBlock.get();
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-211617-6449/analysis/acceptance-project/src/customerimporter/cuit.py
from customerimporter.identification import Identification


class Cuit(Identification):

    # type

    @classmethod
    def type_code(cls):
        return "C"

    # validation

    @classmethod
    def assert_valid(cls, a_number):
        # "23-25666777-9" size 13
        cls._assert_size_between(a_number, cls.minimum_size(), cls.maximum_size())
        cls._assert_has_separators(a_number)
        cls._assert_has_valid_prefix(a_number)
        cls._assert_all_digits(cls.verification_digit_of(a_number))
        cls._assert_all_digits(cls.body_of(a_number))

    @classmethod
    def maximum_size(cls):
        return 13

    @classmethod
    def minimum_size(cls):
        return 12

    @classmethod
    def separator(cls):
        return "-"

    @classmethod
    def valid_prefixes(cls):
        return ["20", "23", "24", "25", "26", "27", "30", "33", "34"]

    # validation - private

    @classmethod
    def _assert_has_separators(cls, a_number):
        if a_number[2] != cls.separator() or a_number[-2] != cls.separator(): cls._signal_invalid_value()

    @classmethod
    def _assert_has_valid_prefix(cls, a_number):
        if cls.prefix_of(a_number) not in cls.valid_prefixes(): cls._signal_invalid_value()

    # error messages

    @classmethod
    def invalid_value_error_description(cls):
        return "Invalid CUIT number"

    # number

    @classmethod
    def body_of(cls, a_number):
        return a_number[3:-2]

    @classmethod
    def prefix_of(cls, a_number):
        return a_number[:2]

    @classmethod
    def verification_digit_of(cls, a_number):
        return a_number[-1:]

    def cuit_number_if_none(self, a_none_block):
        return self.number()

    # testing

    def is_cuit(self):
        return True
